src/auth/auth.cpp

```diff
@@ -232,72 +275,156 @@ auto ParseJson(std::string_view str) {
+  if (!modules_.contains("basic")) {
     /*
-     * MODULE AUTH STORAGE
+     * LOCAL AUTH STORAGE
      */
-    const auto license_check_result = license::global_license_checker.IsEnterpriseValid(utils::global_settings);
-    if (license_check_result.HasError()) {
-      spdlog::warn(license::LicenseCheckErrorToString(license_check_result.GetError(), "authentication modules"));
+    auto user = GetUser(username);
+    if (!user) {
+      spdlog::warn(utils::MessageWithLink("Couldn't authenticate user '{}' because the user doesn't exist.", username,
+                                          "https://memgr.ph/auth"));
+      return std::nullopt;
+    }
+    if (!user->CheckPassword(password)) {
+      spdlog::warn(utils::MessageWithLink("Couldn't authenticate user '{}' because the password is not correct.",
+                                          username, "https://memgr.ph/auth"));
       return std::nullopt;
     }
+    if (user->UpgradeHash(password)) {
+      SaveUser(*user);
+    }
 
-    nlohmann::json params = nlohmann::json::object();
-    params["username"] = username;
-    params["password"] = password;
+    return user;
+  }
 
-    auto ret = module_.Call(params, FLAGS_auth_module_timeout_ms);
+  if (!HasAuthModulePrerequisites("basic")) {
+    return std::nullopt;
+  }
 
-    // Verify response integrity.
-    if (!ret.is_object() || ret.find("authenticated") == ret.end() || ret.find("role") == ret.end()) {
-      return std::nullopt;
-    }
-    const auto &ret_authenticated = ret.at("authenticated");
-    const auto &ret_role = ret.at("role");
-    if (!ret_authenticated.is_boolean() || !ret_role.is_string()) {
+  nlohmann::json params = nlohmann::json::object();
+  params["username"] = username;
+  params["password"] = password;
+
+  auto ret = modules_.at("basic").Call(params, FLAGS_auth_module_timeout_ms);
+
+  // Verify response integrity.
+  if (!ret.is_object() || ret.find("authenticated") == ret.end() || ret.find("role") == ret.end()) {
+    return std::nullopt;
+  }
+  const auto &ret_authenticated = ret.at("authenticated");
+  const auto &ret_role = ret.at("role");
+  if (!ret_authenticated.is_boolean() || !ret_role.is_string()) {
+    return std::nullopt;
+  }
+  auto is_authenticated = ret_authenticated.get<bool>();
+  if (!is_authenticated) return std::nullopt;
+  const auto &rolename = ret_role.get<std::string>();
+
+  auto role = GetRole(rolename);
+  if (!role) {
+    spdlog::warn(utils::MessageWithLink("Couldn't authenticate external user because the role {} doesn't exist.",
+                                        rolename, "https://memgr.ph/auth"));
+    return std::nullopt;
+  }
+  auto already_existing_user = GetUser(username);
+  if (already_existing_user) {
+    spdlog::warn(utils::MessageWithLink("Couldn't authenticate external user because a local user {} already exists.",
+                                        rolename, "https://memgr.ph/auth"));
+    return std::nullopt;
+  }
```

Comment:
Why should we care about local users?
If we allow only LDAP or native auth, the two logics should not mix at all.
If we allow both, we need to be able to handle this case (not simply return an error).

Reply:
From today’s Ante <-> Andrej call: login should fail if a SSO username has a duplicate local (native auth) username; this is in order to avoid audit log confusion